x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -544,4 +630,52 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
         LOGGER.debug("Optimized physical plan:\n{}", plan);
         return plan;
     }
+
+    class IndexResolutionWrappingListener implements ActionListener<EnrichResolution> {
+
+        private final ActionListener<Tuple<EnrichResolution, IndexResolution>> delegate;
+        private final IndexResolution indexResolution;
+
+        IndexResolutionWrappingListener(
+            ActionListener<Tuple<EnrichResolution, IndexResolution>> delegate,
+            IndexResolution indexResolution
+        ) {
+            this.delegate = delegate;
+            this.indexResolution = indexResolution;
+        }
+
+        @Override
+        public void onResponse(EnrichResolution enrichResolution) {
+            delegate.onResponse(new Tuple<>(enrichResolution, indexResolution));
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            delegate.onFailure(e);
+        }
+    }
+
+    class EnrichResolutionWrappingListener implements ActionListener<IndexResolution> {
+
```

Comment:
These two classes are very similar - use a base class with Generics to pass the parameter in.